Foundation/NSError.swift

```diff
@@ -187,29 +196,218 @@ extension CFError : _NSBridgable {
+    }
+}
+
+/// Describes an error type that specifically provides a domain, code,
+/// and user-info dictionary.
+public protocol CustomNSError : Error {
+    /// The domain of the error.
+    static var errorDomain: String { get }
 
-public protocol _ObjectTypeBridgeableErrorType : Swift.Error {
+    /// The error code within the given domain.
+    var errorCode: Int { get }
+
+    /// The user-info dictionary.
+    var errorUserInfo: [String : Any] { get }
+}
+
+public extension Error where Self : CustomNSError {
+    /// Default implementation for customized NSErrors.
+    var _domain: String { return Self.errorDomain }
+
+    /// Default implementation for customized NSErrors.
+    var _code: Int { return self.errorCode }
+}
+
+public extension Error {
+    /// Retrieve the localized description for this error.
+    var localizedDescription: String {
+        return NSError(domain: _domain, code: _code, userInfo: nil).localizedDescription
+    }
+}
+
+/// Retrieve the default userInfo dictionary for a given error.
+public func _swift_Foundation_getErrorDefaultUserInfo(_ error: Error) -> Any? {
+    // TODO: Implement info value providers and return the code that was deleted here.
+    let hasUserInfoValueProvider = false
+
+    // Populate the user-info dictionary
+    var result: [String : Any]
+
+    // Initialize with custom user-info.
+    if let customNSError = error as? CustomNSError {
+        result = customNSError.errorUserInfo
+    } else {
+        result = [:]
+    }
+
+    // Handle localized errors. If we registered a user-info value
+    // provider, these will computed lazily.
+    if !hasUserInfoValueProvider,
+    let localizedError = error as? LocalizedError {
+        if let description = localizedError.errorDescription {
+            result[NSLocalizedDescriptionKey] = description
+        }
+
+        if let reason = localizedError.failureReason {
+            result[NSLocalizedFailureReasonErrorKey] = reason
+        }
+
+        if let suggestion = localizedError.recoverySuggestion {
+            result[NSLocalizedRecoverySuggestionErrorKey] = suggestion
+        }
+
+        if let helpAnchor = localizedError.helpAnchor {
+            result[NSHelpAnchorErrorKey] = helpAnchor
+        }
+    }
+
+    // Handle recoverable errors. If we registered a user-info value
+    // provider, these will computed lazily.
+    if !hasUserInfoValueProvider,
+    let recoverableError = error as? RecoverableError {
+        result[NSLocalizedRecoveryOptionsErrorKey] =
+        recoverableError.recoveryOptions
+        result[NSRecoveryAttempterErrorKey] = _NSErrorRecoveryAttempter()
+    }
+
+    return result
+}
+
+// NSError and CFError conform to the standard Error protocol. Compiler
+// magic allows this to be done as a "toll-free" conversion when an NSError
+// or CFError is used as an Error existential.
+
+extension NSError {
+    /// The "embedded" NSError is itself.
+    public func _getEmbeddedNSError() -> AnyObject? {
```

Comment:
`_getEmbeddedNSError` isn't actually used when we don't have an Objective-C runtime, so you can remove this bit.